src/Kernel/client/visualizer.ts

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { createExecutionPathVisualizer, Circuit, StyleConfig, Operation } from "./ExecutionPathVisualizer";
+
+type GateRegistry = {
+    [id: string]: Operation,
+};
+
+// Flag to ensure that we only inject custom JS into browser once
+let isScriptInjected = false;
+
+// Event handler to visually signal to user that the gate can be zoomed out on ctrl-click
+window.addEventListener('keydown', (ev) => {
+    if (ev.key !== 'Control') return;
+    document.querySelectorAll('[data-zoom-out="true"]').forEach((el: HTMLElement) => {
+        el.style.cursor = 'zoom-out';
+    });
+});
+
+// Event handler to visually signal to user that the gate can be zoomed in on click
+window.addEventListener('keyup', (ev) => {
+    if (ev.key !== 'Control') return;
+    document.querySelectorAll('[data-zoom-out="true"]').forEach((el: HTMLElement) => {
+        el.style.cursor = 'default';
+    });
+    document.querySelectorAll('[data-zoom-in="true"]').forEach((el: HTMLElement) => {
+        el.style.cursor = 'zoom-in';
+    });
+});
+
+export class Visualizer {
+    userStyleConfig: StyleConfig = {};
+    displayedCircuit: Circuit = null;
+    id: string = "";
+    gateRegistry: GateRegistry = {};
+
+    constructor(id: string, userStyleConfig: StyleConfig) {
+        this.id = id;
+        this.userStyleConfig = userStyleConfig;
+    }
+
+    visualize(circuit: Circuit, renderDepth: number = 0) {
+        // Assign unique IDs to each operation
+        circuit.operations.forEach((op, i) => this.fillGateRegistry(op, i.toString()));
+
+        // Render operations at starting at given depth
+        circuit.operations = this.selectOpsAtDepth(circuit.operations, renderDepth);
+        this.renderCircuit(circuit);
+    }
+
+    // Depth-first traversal to assign unique ID to operation.
+    // The operation is assigned the id `id` and its `i`th child is recursively given
+    // the id `${id}-${i}`.
+    fillGateRegistry(operation: Operation, id: string) {
+        if (operation.dataAttributes == null) operation.dataAttributes = {};
+        operation.dataAttributes["id"] = id;
```

Comment:
[nit] Just for consistency...
```suggestion
        operation.dataAttributes['id'] = id;
```